android/src/main/java/io/ably/flutter/plugin/AblyEventStreamHandler.java

```
@@ -0,0 +1,107 @@
+        public void onChannelStateChanged(io.ably.lib.realtime.ChannelStateListener.ChannelStateChange stateChange){
+            eventSink.success(stateChange);
+        }
+    }
+
+    @Override
+    public void onListen(Object object, EventChannel.EventSink uiThreadEventSink) {
+        MainThreadEventSink eventSink = new MainThreadEventSink(uiThreadEventSink);
+        methodCallHandler.<AblyFlutterMessage<String>>ablyDo((AblyFlutterMessage)object, (ablyLibrary, message) -> {
+            String eventName = message.message;
+            switch(eventName) {
+                case "realtime_onConnectionStateChanged":
+                    connectionStateListener = new PluginConnectionStateListener(eventSink);
+                    ablyLibrary.getRealtime(message.handle).connection.on(connectionStateListener);
+                    return;
+                case "realtime_onChannelStateChanged":
+                    channelStateListener = new PluginChannelStateListener(eventSink);
+                    // ablyLibrary.getRealtime(message.handle).connection.on(channelStateListener);
+                    return;
+                default:
+                    eventSink.error("unhandled event", null, null);
+            }
+        });
+    }
+
+    @Override
+    public void onCancel(Object object) {
+        if(object==null){
+            System.out.println("Cannot process null input on cancel");
+            return;
+        }
+        methodCallHandler.<AblyFlutterMessage<String>>ablyDo((AblyFlutterMessage)object, (ablyLibrary, message) -> {
+            String eventName = message.message;
+            switch (eventName) {
+                case "realtime_onConnectionStateChanged":
+                    ablyLibrary.getRealtime(message.handle).connection.off(connectionStateListener);
+                case "realtime_onChannelStateChanged":
+                    // ablyLibrary.getRealtime(handle).connection.off(connectionStateListener);
```

Why is this commented code?

---

@mattheworiordan it is a WIP, so is commented out.

---

Perhaps preferable to raise an exception or trigger an assertion failure here so it fails early at runtime.

---

@QuintinWillison raising an exception here just kills the app.
`onCancel` method can be called only if a listener is instantiated.

So, onListen is updated to throw runtime exception in d59f89a2e4a18ccdcef6443f45d90ac453507501

---

@tiholic [as above](https://github.com/ably/ably-flutter/pull/6#discussion_r425642598) I've commented on that commit.